### PR TITLE
Implement `splitdef`/`combinedef`

### LINKED ISSUE #58: Mocking does not support Nothing kwarg defaults
```
julia [invenia]> p = @patch function f(x; y=nothing)
                     (x, x, y, y)
                 end
ERROR: MethodError: no method matching binding_expr(::Void)
Closest candidates are:
  binding_expr(::Module) at /Users/ericdavies/.playground/share/invenia/packages/v0.6/Mocking/src/expr.jl:29
  binding_expr(::Union) at /Users/ericdavies/.playground/share/invenia/packages/v0.6/Mocking/src/expr.jl:36
  binding_expr(::Type) at /Users/ericdavies/.playground/share/invenia/packages/v0.6/Mocking/src/expr.jl:32
  ...
Stacktrace:
 [1] adjust_bindings(::Dict{Symbol,Void}) at /Users/ericdavies/.playground/share/invenia/packages/v0.6/Mocking/src/expr.jl:56
 [2] Mocking.Patch(::Expr, ::Function, ::Dict{Symbol,Void}) at /Users/ericdavies/.playground/share/invenia/packages/v0.6/Mocking/src/Mocking.jl:48
```

### LINKED ISSUE #61: patching with where
There are some things that break the `@patch` macro.

They include

```
julia> @patch foo(::Type{T}) where {T<:AbstractArray{S}} where S = S
ERROR: LoadError: ArgumentError: expression is not a function definition
Stacktrace:
 [1] @patch(::LineNumberNode, ::Module, ::Expr) at /Users/oxinabox/.julia/packages/Mocking/1Wk5l/src/Mocking.jl:119
in expression starting at REPL[6]:1




julia> @patch foo(::Type{T}) where {T<:AbstractArray} = T

ERROR: LoadError: ArgumentError: expression is not a function definition
Stacktrace:
 [1] @patch(::LineNumberNode, ::Module, ::Expr) at /Users/oxinabox/.julia/packages/Mocking/1Wk5l/src/Mocking.jl:119
in expression starting at REPL[7]:1


julia> @patch function foo(::Type{T}) where {T<:AbstractArray}
       T
       end
ERROR: LoadError: expression is not a valid function call: foo(::Type{T})
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] ingest_signature!(::Mocking.Bindings, ::Expr) at /Users/oxinabox/.julia/packages/Mocking/1Wk5l/src/bindings.jl:183
 [3] Mocking.Bindings(::Expr) at /Users/oxinabox/.julia/packages/Mocking/1Wk5l/src/bindings.jl:27
 [4] @patch(::LineNumberNode, ::Module, ::Expr) at /Users/oxinabox/.julia/packages/Mocking/1Wk5l/src/Mocking.jl:125
in expression starting at REPL[11]:1
```

etc.

All these are valid without the `@patch` macro.

### PR DESCRIPTION
**Pull request by omus**
After looking over the usage of MacroTools in #60 I found that the proposed implementation wasn't handling the various function definitions that can occur. I found the `splitdef` and `combinedef` implementations provided by MacroTools but I discovered that those functions were not fully tested and were quite slow. Additionally, the extra dependences MacroTools added seemed unnecessary here.

```julia
julia> using BenchmarkTools, MacroTools, Mocking

julia> @btime MacroTools.splitdef(:(f(x) = 1))
  346.797 μs (543 allocations: 35.33 KiB)
Dict{Symbol,Any} with 5 entries:
  :name        => :f
  :args        => Any[:x]
  :kwargs      => Any[]
  :body        => quote…
  :whereparams => ()

julia> @btime Mocking.splitdef(:(f(x) = 1))
  810.213 ns (14 allocations: 1.14 KiB)
Dict{Symbol,Any} with 4 entries:
  :type => :(=)
  :args => Any[:x]
  :body => quote…
  :name => :f
```

I think it would be good to make these functions into a separate package for general usage but for now they can live in Mocking (I'm not sure what a good name for this package would be).

- Fixes #61
- Fixes #58

### CODE REVIEW

File: src/expr.jl [hunk 1]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+- `:rtype`: Return type of the function
+- `:whereparams`: Where parameters
+- `:body`: Function body (not present for empty functions)
+
+All parts are optional with the exception of `:type`. When `throw=true` and the provided
+expression is not a function definition than an exception will be raised. When `throw=false`
+then `nothing` will be returned for invalid definitions.
+"""
+function splitdef(ex::Expr; throw::Bool=true)
+    def = Dict{Symbol,Any}()
```
Comment (iamed2) [inline src/expr.jl]: Why not a type or a NamedTuple?

Comment (omus) [reply src/expr.jl]: I would be good with returning a type. I ended up using a `Dict` to mostly keep compatibly with MacroTools' version

File: test/expr.jl [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+
+macro test_splitdef_invalid(expr)
+    result = quote
+        @test_throws ArgumentError splitdef($expr)
+        @test splitdef($expr, throw=false) === nothing
+    end
+    return esc(result)
+end
+
+@testset "splitdef / combinedef" begin
```
Comment (rofinn) [inline test/expr.jl]: If the objective is to keep the same API as MacroTools then it might be good to make MacroTools a test dependency and also compare all the cases that should be the same and explicitly show which cases MacroTools won't handle.

Comment (omus) [reply test/expr.jl]: I'm not sure that would be worth the effort. The functions here definitely don't have the same API (note `:type` is never returned by MacroTools.splitdef). I'll update the description to mention what isn't handled by MacroTools

**Comment by omus** [comment]
I'll try to put this code into a a separate package. I think ExprTools.jl is a reasonable name. My reason for not putting back into MacroTools.jl is so that we don't have to add unnecessary dependencies here.

File: src/Mocking.jl [hunk 3]
```diff
@@ -40,17 +40,17 @@ struct Patch{T}
...
     # Expect a named function in long-form or short-form
-    if expr.head === :function || (expr.head === :(=) && expr.args[1].head == :call)
-        target_name = expr.args[1].args[1]
-        params = expr.args[1].args[2:end]
-        body = expr.args[2]
-    elseif expr.head == :(->)
-        throw(ArgumentError("expression needs to be a named function"))
-    else
+    if def === nothing
         throw(ArgumentError("expression is not a function definition"))
```
Comment (oxinabox) [inline src/Mocking.jl]: Why not make `splitdef` throw the error?

File: src/Mocking.jl [hunk 4]
```diff
@@ -40,17 +40,17 @@ struct Patch{T}
...
+    elseif def[:type] == :(->)
```
Comment (oxinabox) [inline src/Mocking.jl]: can we not use the word `type` for something not releated to the typesytem?
Maybe  `kind`, or `form` ?

Also should we detect this error condition rather via

```suggestion
    elseif !haskey(def, :name)
```

Since that is what we are after?

File: src/Mocking.jl [hunk 5]
```diff
@@ -59,12 +59,12 @@ macro patch(expr::Expr)
         string(target_name)
     end
 
-    # Need to evaluate the body of the function in the context of the `@patch` macro in
-    # order to support closures.
-    # func = Expr(:(->), Expr(:tuple, params...), body)
-    func = Expr(:(=), Expr(:call, gensym(patch_name), params...), body)
+    def[:name] = gensym(string(patch_name, "_patch"))
```
Comment (oxinabox) [inline src/Mocking.jl]: Rather than the above stuff to normalize the different forms a `target_name` might have depending on if it is a `Symbol` or an `Expr`
we can just do:

`def[:name] = gensym(Symbol(target_name, :_patch))`

I think that has literally the exact same effect.

demo
```
julia> gensym(Symbol(:foo, :_patch))
Symbol("##foo_patch#400")

julia> gensym(Symbol(:(Base.foo), :_patch))
Symbol("##Base.foo_patch#401")
```

File: src/expr.jl [hunk 6]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+            msg = "Function definition contains $section\n$(sprint(Meta.dump, ex))"
+            Base.throw(ArgumentError(msg))
+        else
+            nothing
+        end
+    end
+
+    # long-form `:function`, short-form `:(=)`, anonymous `:(->)`
+    if !(ex.head === :function || ex.head === :(=) || ex.head === :(->))
+        return invalid_def("invalid function type `$(repr(ex.head))`")
```
Comment (oxinabox) [inline src/expr.jl]: again would rather use some word other than `type`

I think `form` sounds good
```suggestion
        return invalid_def("invalid function form `$(repr(ex.head))`")
```

File: src/expr.jl [hunk 7]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+    end
+
+    def[:type] = ex.head
+    anon = ex.head == :(->)
```
Comment (oxinabox) [inline src/expr.jl]: One can also create an anon function using

```

julia> function()
       10^2
       end
#5 (generic function with 1 method)
```

I do not see a test for it below

Comment (oxinabox) [reply src/expr.jl]: Another fun test case is

```
function(::T) where T
       10^2
end
```

File: src/expr.jl [hunk 8]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+        Expr(:curly, def[:name], def[:params]...)
+    elseif haskey(def, :name)
+        def[:name]
+    else
+        nothing
+    end
+
+    # Empty generic function definitions must not contain additional keys
+    if !haskey(def, :body) && any(k -> haskey(def, k), [:args, :kwargs, :rtype, :whereparams])
+        extra = [:args, :kwargs, :rtype, :whereparams]
```
Comment (oxinabox) [inline src/expr.jl]:
```suggestion
        extra = (:args, :kwargs, :rtype, :whereparams)
```

File: src/expr.jl [hunk 9]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+    name = if haskey(def, :params)
+        Expr(:curly, def[:name], def[:params]...)
+    elseif haskey(def, :name)
+        def[:name]
+    else
+        nothing
+    end
+
+    # Empty generic function definitions must not contain additional keys
+    if !haskey(def, :body) && any(k -> haskey(def, k), [:args, :kwargs, :rtype, :whereparams])
```
Comment (oxinabox) [inline src/expr.jl]:
```suggestion
    if !haskey(def, :body) && any(k -> haskey(def, k), (:args, :kwargs, :rtype, :whereparams))
```

File: src/expr.jl [hunk 10]
```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
...
+        throw(ArgumentError(string(
+            "Function definitions without a body must not contain keys: ",
+            join(string.('`', repr.(setdiff(extra, keys(def))), '`'), ", ", ", or "),
+        )))
+    end
+
+    # Combine args and kwargs
+    args = Any[]
+    !isempty(get(def, :kwargs, [])) && push!(args, Expr(:parameters, def[:kwargs]...))
+    !isempty(get(def, :args, [])) && append!(args, def[:args])
```
Comment (oxinabox) [inline src/expr.jl]: For all in intents and purposes  the above and

```suggestion
    haskey(def, :args)) && append!(args, def[:args])
```

or even 

```suggestion
    append!(args, get(def, :args, []))
```

are going to be the same.

And I think they are clearer than 
```
!isempty(get(def, :args, [])) && append!(args, def[:args])
```

File: test/expr.jl [hunk 11]
```diff
@@ -0,0 +1,538 @@
+macro audit(expr::Expr)
```
Comment (oxinabox) [inline test/expr.jl]: A comment saying what this does would be nice.

File: src/Mocking.jl (hunk 3 above)
Comment (omus) [reply src/Mocking.jl]: I think originally I didn't have the `splitdef` exception throwing working very well. At this point I think I can just have `splitdef` throw the exception.

Comment (omus) [reply src/Mocking.jl]: Oh, I may have been playing with the idea of returning the exception and letting it fail at runtime instead of during macro expansion

File: src/Mocking.jl (hunk 4 above)
Comment (omus) [reply src/Mocking.jl]: I think `:form` could be a good alternative. I agree `:type` is rather over used. Hmm, at first I didn't like the `haskey` idea but it's growing on me fast.

File: src/Mocking.jl (hunk 5 above)
Comment (omus) [reply src/Mocking.jl]: Good point. The normalizing is just a nicety but it's probably not work the extra complexity.

Comment (oxinabox) [reply src/Mocking.jl]: I think the normalizing even ends up like this? No?

File: src/expr.jl (hunk 10 above)
Comment (omus) [reply src/expr.jl]: After reading this I have been adding some tests for `(;) -> nothing` which seems to be a corner case I wasn't previously testing for.

File: src/Mocking.jl (hunk 5 above)
Comment (omus) [reply src/Mocking.jl]: I should be able to drop the extra string calls like you suggested. I was also thinking you were suggesting dropping normalization all together (e.g. `:(Base.foo)` → `:foo`). Reading your original comment again I don't think you were suggesting that but I may still do that as the main point to adjusting the gensym'd name is to improve readability in the stack traces.

File: src/Mocking.jl (hunk 4 above)
Comment (iamed2) [reply src/Mocking.jl]: I'm supportive of calling it `head` if it means `expr.head`

File: src/Mocking.jl (hunk 5 above)
Comment (oxinabox) [reply src/Mocking.jl]: the current code doesn't normalize `:(Base.foo)` → `:foo)`
AFAICT
All it does is call `string(:(Base.foo))` which just gives `"Base.foo"` ?

File: src/Mocking.jl (hunk 4 above)
Comment (omus) [reply src/Mocking.jl]: > can we not use the word `type` for something not releated to the typesytem?
Maybe `kind`, or `form` ?

I think a different word is a good idea. I was going to go with `:form` but I like to use that to indicate long-form/short-form definition syntax. Since this is always the function definition head I think `:head` is quite reasonable.

File: src/Mocking.jl (hunk 5 above)
Comment (omus) [reply src/Mocking.jl]: This is the normalization code:
```julia
patch_name = if target_name isa Symbol  # f(...)
    target_name
elseif target_name.head === :.  # Base.f(...)
    target_name.args[2].value
else
    string(target_name)
end
```

File: test/expr.jl (hunk 11 above)
Comment (omus) [reply test/expr.jl]: I added a docstring

File: src/expr.jl (hunk 7 above)
Comment (omus) [reply src/expr.jl]: I'm now testing against long-form anonymous functions

File: src/expr.jl (hunk 10 above)
Comment (omus) [reply src/expr.jl]: `haskey` works here.

File: src/expr.jl (hunk 6 above)
Comment (omus) [reply src/expr.jl]: I've updated this to say "head"

File: src/Mocking.jl (hunk 5 above)
Comment (omus) [reply src/Mocking.jl]: I refactored the normalization code and the patch name creation. Let me know what you think.

Comment (oxinabox) [reply src/Mocking.jl]: Oh I some how skimmed over the  `elseif` bit.
New code is fine.

File: src/Mocking.jl [hunk 12]
```diff
@@ -40,31 +40,31 @@ struct Patch{T}
...
     end
 
-    patch_name = if target_name isa Symbol  # f(...)
-        target_name
-    elseif target_name.head === :.  # Base.f(...)
-        target_name.args[2].value
+    # Include the target function name in the patch to make stack traces easier to read.
+    # If the provided target uses a fully-qualified reference we'll just extract the name
+    # of the function (e.g `Base.foo` -> `foo`).
+    target_name = if target isa Expr && target.head === :.
```
Comment (oxinabox) [inline src/Mocking.jl]: How do you feel about

```suggestion
    target_name = if  Meta.isexpr(target, :.)
```

File: test/expr.jl [hunk 13]
```diff
@@ -0,0 +1,805 @@
+"""
+    @audit expr -> Tuple{Any,Expr}
+
+Evaluate the expression and return both the result and the original expression. Useful for
+ensuring that the provided expression is syntactically valid. If provided expression cannot
+be evaulated the exception will be returned instead of the result.
```
Comment (oxinabox) [inline test/expr.jl]:
```suggestion
be evaluated the exception will be returned instead of the result.
```

File: test/expr.jl [hunk 14]
```diff
@@ -0,0 +1,805 @@
...
+                :body => quote
+                    x
+                end
+            )
+            c_expr = combinedef(d)
+            @test strip_lineno!(c_expr) == strip_lineno!(expr)
+        end
+    end
+
+    @testset "invalid definitions" begin
```
Comment (oxinabox) [inline test/expr.jl]: Maybe add
```
 (a::Number::Int)->a
```
